File: Python/examples/base64_utils_example.py
"""
AllToolkit - Python Base64 Utilities Example

Demonstrates usage of the Base64 encoding/decoding utilities.

Run with: python base64_utils_example.py
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'base64_utils'))

from mod import Base64Utils


def example_basic_encoding():
    """Example 1: Basic encoding and decoding."""
    print("\n" + "=" * 60)
    print("Example 1: Basic Encoding and Decoding")
    print("=" * 60)
    
    # Encode a string
    original = "Hello, World!"
    encoded = Base64Utils.encode(original)
    print(f"Original: {original}")
    print(f"Encoded:  {encoded}")
    
    # Decode back
    decoded = Base64Utils.decode(encoded)
    print(f"Decoded:  {decoded}")
    print(f"Match:    {original == decoded}")


def example_binary_data():
    """Example 2: Working with binary data."""
    print("\n" + "=" * 60)
    print("Example 2: Binary Data Encoding")
    print("=" * 60)
    
    # Binary data (could be image bytes, etc.)
    binary_data = bytes([0x00, 0x01, 0x02, 0xFF, 0xFE, 0xFD])
    print(f"Binary data: {binary_data}")
    
    # Encode to Base64
    encoded = Base64Utils.encode(binary_data)
    print(f"Encoded:     {encoded}")
    
    # Decode back to bytes
    decoded = Base64Utils.decode_to_bytes(encoded)
    print(f"Decoded:     {decoded}")
    print(f"Match:       {binary_data == decoded}")


def example_urlsafe():
    """Example 3: URL-safe Base64 for URLs and filenames."""
    print("\n" + "=" * 60)
    print("Example 3: URL-safe Base64 (RFC 4648)")
    print("=" * 60)
    
    # Data that produces + and / in standard Base64
    data = "hello+world/test>data"
    
    standard = Base64Utils.encode(data)
    urlsafe = Base64Utils.encode_urlsafe(data)
    urlsafe_no_pad = Base64Utils.encode_urlsafe(data, padding=False)
    
    print(f"Original:           {data}")
    print(f"Standard Base64:    {standard}")
    print(f"URL-safe Base64:    {urlsafe}")
    print(f"URL-safe (no pad):  {urlsafe_no_pad}")
    
    # Decode URL-safe
    decoded = Base64Utils.decode_urlsafe(urlsafe)
    print(f"Decoded:            {decoded}")
    print(f"Match:              {data == decoded}")


def example_conversion():
    """Example 4: Converting between standard and URL-safe."""
    print("\n" + "=" * 60)
    print("Example 4: Converting Between Formats")
    print("=" * 60)
    
    # Standard Base64 with + and /
    standard = "aGVsbG8+/test="
    print(f"Standard Base64:    {standard}")
    
    # Convert to URL-safe
    urlsafe = Base64Utils.to_urlsafe(standard)
    print(f"URL-safe:           {urlsafe}")
    
    # Convert back to standard
    back_to_standard = Base64Utils.from_urlsafe(urlsafe)
    print(f"Back to standard:   {back_to_standard}")
    print(f"Match:              {standard == back_to_standard}")


def example_validation():
    """Example 5: Validating Base64 strings."""
    print("\n" + "=" * 60)
    print("Example 5: Base64 Validation")
    print("=" * 60)
    
    valid_strings = [
        "SGVsbG8=",
        "YWJj",
        "",
    ]
    
    invalid_strings = [
        "Invalid!!!",
        "Not@Base64#",
    ]
    
    print("Valid strings:")
    for s in valid_strings:
        result = Base64Utils.is_valid(s)
        print(f"  '{s}' -> {result}")
    
    print("\nInvalid strings:")
    for s in invalid_strings:
        result = Base64Utils.is_valid(s)
        print(f"  '{s}' -> {result}")
    
    # URL-safe validation
    print("\nURL-safe validation:")
    urlsafe_valid = "aGVsbG8-"
    urlsafe_invalid = "aGVsbG8+"
    print(f"  '{urlsafe_valid}' (urlsafe=True) -> {Base64Utils.is_valid(urlsafe_valid, urlsafe=True)}")
    print(f"  '{urlsafe_invalid}' (urlsafe=True) -> {Base64Utils.is_valid(urlsafe_invalid, urlsafe=True)}")


def example_length_calculation():
    """Example 6: Calculating encoded/decoded lengths."""
    print("\n" + "=" * 60)
    print("Example 6: Length Calculations")
    print("=" * 60)
    
    test_lengths = [1, 2, 3, 10, 100, 1000]
    
    print(f"{'Input Length':<15} {'Encoded':<10} {'Encoded (no pad)':<20}")
    print("-" * 50)
    
    for length in test_lengths:
        encoded_with_pad = Base64Utils.encoded_length(length)
        encoded_no_pad = Base64Utils.encoded_length(length, padding=False)
        print(f"{length:<15} {encoded_with_pad:<10} {encoded_no_pad:<20}")
    
    print("\nMax decoded length from Base64:")
    base64_lengths = [4, 8, 100, 136]
    for bl in base64_lengths:
        max_decoded = Base64Utils.decoded_max_length(bl)
        print(f"  Base64 length {bl} -> max {max_decoded} bytes")


def example_unicode():
    """Example 7: Unicode and multi-byte characters."""
    print("\n" + "=" * 60)
    print("Example 7: Unicode Support")
    print("=" * 60)
    
    test_strings = [
        "Hello, 世界!",
        "🎉 Emoji support 🚀",
        "日本語テキスト",
        "العربية",
    ]
    
    for original in test_strings:
        encoded = Base64Utils.encode(original)
        decoded = Base64Utils.decode(encoded)
        print(f"Original: {original}")
        print(f"Encoded:  {encoded}")
        print(f"Decoded:  {decoded}")
        print(f"Match:    {original == decoded}\n")


def example_practical_use_cases():
    """Example 8: Practical use cases."""
    print("\n" + "=" * 60)
    print("Example 8: Practical Use Cases")
    print("=" * 60)
    
    # Use case 1: Encoding data for URL parameters
    print("\n1. URL Parameter Encoding:")
    user_data = "user+name@example.com"
    url_safe = Base64Utils.encode_urlsafe(user_data, padding=False)
    print(f"   Data: {user_data}")
    print(f"   URL-safe: {url_safe}")
    
    # Use case 2: Encoding binary data for JSON
    print("\n2. Binary Data for JSON:")
    binary = bytes([0x89, 0x50, 0x4E, 0x47])  # PNG file header
    json_safe = Base64Utils.encode(binary)
    print(f"   Binary: {binary}")
    print(f"   JSON-safe: {json_safe}")
    
    # Use case 3: Simple obfuscation
    print("\n3. Simple Data Obfuscation:")
    sensitive = "secret_token_12345"
    obfuscated = Base64Utils.encode(sensitive)
    print(f"   Original:    {sensitive}")
    print(f"   Obfuscated:  {obfuscated}")
    print(f"   Deobfuscated: {Base64Utils.decode(obfuscated)}")


def example_convenience_functions():
    """Example 9: Using convenience functions."""
    print("\n" + "=" * 60)
    print("Example 9: Convenience Functions")
    print("=" * 60)
    
    # Import convenience functions directly
    from mod import encode, decode, encode_urlsafe, is_valid
    
    # Use without class prefix
    original = "Quick test"
    encoded = encode(original)
    decoded = decode(encoded)
    
    print(f"Original: {original}")
    print(f"Encoded:  {encoded}")
    print(f"Decoded:  {decoded}")
    print(f"Valid:    {is_valid(encoded)}")


def main():
    """Run all examples."""
    print("\n" + "=" * 60)
    print("AllToolkit - Python Base64 Utilities Examples")
    print("=" * 60)
    
    examples = [
        example_basic_encoding,
        example_binary_data,
        example_urlsafe,
        example_conversion,
        example_validation,
        example_length_calculation,
        example_unicode,
        example_practical_use_cases,
        example_convenience_functions,
    ]
    
    for example in examples:
        try:
            example()
        except Exception as e:
            print(f"\nError in {example.__name__}: {e}")
    
    print("\n" + "=" * 60)
    print("All examples completed!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: C#/examples/base64_utils_example.cs
using System;
using System.Text;
using AllToolkit;

namespace AllToolkit.Examples
{
    /// <summary>
    /// Example usage of Base64Utils class.
    /// </summary>
    class Base64UtilsExample
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Base64Utils Examples");
            Console.WriteLine("====================");
            Console.WriteLine();

            Example1_BasicEncoding();
            Example2_UrlSafeEncoding();
            Example3_BinaryData();
            Example4_Validation();
            Example5_ErrorHandling();
            Example6_LengthCalculations();

            Console.WriteLine();
            Console.WriteLine("All examples completed!");
        }

        static void Example1_BasicEncoding()
        {
            Console.WriteLine("Example 1: Basic Encoding and Decoding");
            Console.WriteLine("--------------------------------------");

            // Basic encoding
            string original = "Hello, World!";
            string encoded = Base64Utils.Encode(original);
            Console.WriteLine($"Original: {original}");
            Console.WriteLine($"Encoded:  {encoded}");

            // Basic decoding
            string decoded = Base64Utils.Decode(encoded);
            Console.WriteLine($"Decoded:  {decoded}");
            Console.WriteLine();

            // Encoding with different text
            string[] texts = { "A", "AB", "ABC", "ABCD" };
            foreach (string text in texts)
            {
                string enc = Base64Utils.Encode(text);
                Console.WriteLine($"'{text}' -> '{enc}'");
            }
            Console.WriteLine();
        }

        static void Example2_UrlSafeEncoding()
        {
            Console.WriteLine("Example 2: URL-Safe Encoding");
            Console.WriteLine("----------------------------");

            // URL-safe encoding is useful for URLs and filenames
            string original = "user+name/file@example.com";
            
            // Standard Base64 (may contain + and /)
            string standard = Base64Utils.Encode(original);
            Console.WriteLine($"Standard Base64: {standard}");

            // URL-safe Base64 (replaces + with - and / with _)
            string urlSafe = Base64Utils.EncodeUrlSafe(original, padding: true);
            Console.WriteLine($"URL-safe Base64: {urlSafe}");

            // URL-safe without padding (shorter, good for tokens)
            string urlSafeNoPad = Base64Utils.EncodeUrlSafe(original, padding: false);
            Console.WriteLine($"URL-safe (no pad): {urlSafeNoPad}");

            // Decoding URL-safe
            string decoded = Base64Utils.DecodeUrlSafe(urlSafe);
            Console.WriteLine($"Decoded: {decoded}");
            Console.WriteLine();
        }

        static void Example3_BinaryData()
        {
            Console.WriteLine("Example 3: Binary Data Encoding");
            Console.WriteLine("-------------------------------");

            // Encode binary data
            byte[] binaryData = new byte[] { 0x00, 0x01, 0x02, 0x03, 0xFF, 0xFE, 0xFD, 0xFC };
            string encoded = Base64Utils.Encode(binaryData);
            Console.WriteLine($"Binary data length: {binaryData.Length} bytes");
            Console.WriteLine($"Encoded: {encoded}");

            // Decode back to bytes
            byte[] decoded = Base64Utils.DecodeToBytes(encoded);
            Console.WriteLine($"Decoded length: {decoded.Length} bytes");
            Console.WriteLine($"Data matches: {BitConverter.ToString(binaryData) == BitConverter.ToString(decoded)}");
            Console.WriteLine();
        }

        static void Example4_Validation()
        {
            Console.WriteLine("Example 4: Validation");
            Console.WriteLine("---------------------");

            // Valid Base64 strings
            string[] validStrings = {
                "SGVsbG8sIFdvcmxkIQ==",
                "",
                "QQ==",
                "SGVsbG8rV29ybGQvVGVzdA"
            };

            Console.WriteLine("Valid strings:");
            foreach (string s in validStrings)
            {
                bool isValid = Base64Utils.IsValid(s);
                Console.WriteLine($"  '{s}' -> {isValid}");
            }

            // Invalid Base64 strings
            string[] invalidStrings = {
                "NotValid!!!",
                "Hello World",
                "===",
                null
            };

            Console.WriteLine("\nInvalid strings:");
            foreach (string s in invalidStrings)
            {
                bool isValid = Base64Utils.IsValid(s);
                Console.WriteLine($"  '{s}' -> {isValid}");
            }
            Console.WriteLine();
        }

        static void Example5_ErrorHandling()
        {
            Console.WriteLine("Example 5: Error Handling");
            Console.WriteLine("-------------------------");

            // Safe decoding with TryDecode
            string result = Base64Utils.TryDecode("SGVsbG8sIFdvcmxkIQ==");
            Console.WriteLine($"TryDecode valid: '{result}'");

            result = Base64Utils.TryDecode("NotValid!!!");
            Console.WriteLine($"TryDecode invalid: '{result ?? "null"}'");

            result = Base64Utils.TryDecode(null);
            Console.WriteLine($"TryDecode null: '{result ?? "null"}'");

            // Exception-based handling
            try
            {
                Base64Utils.Decode("Invalid!!!");
            }
            catch (FormatException e)
            {
                Console.WriteLine($"Caught FormatException: {e.Message}");
            }

            try
            {
                Base64Utils.Encode((string)null);
            }
            catch (ArgumentNullException e)
            {
                Console.WriteLine($"Caught ArgumentNullException: {e.Message}");
            }
            Console.WriteLine();
        }

        static void Example6_LengthCalculations()
        {
            Console.WriteLine("Example 6: Length Calculations");
            Console.WriteLine("------------------------------");

            // Calculate expected encoded length
            int[] inputLengths = { 1, 2, 3, 10, 100, 1000 };
            
            Console.WriteLine("Input Length -> Encoded Length (with padding):");
            foreach (int len in inputLengths)
            {
                int encodedLen = Base64Utils.GetEncodedLength(len, true);
                Console.WriteLine($"  {len,4} bytes -> {encodedLen,4} chars");
            }

            Console.WriteLine("\nInput Length -> Encoded Length (without padding):");
            foreach (int len in inputLengths)
            {
                int encodedLen = Base64Utils.GetEncodedLength(len, false);
                Console.WriteLine($"  {len,4} bytes -> {encodedLen,4} chars");
            }

            // Calculate max decoded length
            int[] encodedLengths = { 4, 8, 12, 100 };
            Console.WriteLine("\nEncoded Length -> Max Decoded Length:");
            foreach (int len in encodedLengths)
            {
                int maxDecoded = Base64Utils.GetDecodedMaxLength(len);
                Console.WriteLine($"  {len,4} chars -> {maxDecoded,4} bytes");
            }
            Console.WriteLine();
        }
    }
}
